### ISSUE TITLE
conda install --offline does not work

### ISSUE THREAD

**Issue by Zaharid** [conda/conda#2035]
I have conda-built a package called "libnnpdf". Short of a proper fix for #1884 I'd like this to work, instead of the errors that are displayed:

```
 $ conda install libnnpdf --use-local --offline
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ..
Error: No packages found in current linux-64 channels matching: conda

You can search for this package on anaconda.org with

    anaconda search -t conda conda
```

 Appending `--use-index-cache` does not solve the problem.

**Comment by kalefranz** [comment]
https://github.com/ilanschnell/misc/blob/master/conda/c2r.py

**Comment by kalefranz** [comment]
Discussed the issue more with @ilanschnell yesterday, and I think we're iterating toward an awesome solution.

**Comment by Zaharid** [comment]
@kalefranz  Thank you for addressing this.

**Comment by phumke** [comment]
Hey @Zaharid , thanks for bringing this up.  I also need to run in an offline mode, and it was confusing to me to have anaconda claim that all of the packages are pre-bundled but be unable to create a new project from this.

It looks like the way the system is set up, 'root' is a pre-installed environment with all of the packages.  A hacky workaround to install from local would be to add `--clone root` when creating a new environment.  It brings in the world so you'd have to then remove the unwanted packages if you wanted to keep it clean.

`conda create --name your_environment --clone root`

@kalefranz @ilanschnell the solution you guys are working on looks awesome.  I'm excited for the functionality. With root being treated as the local cache for the offline channel, what are your thoughts on having root also get a duplicate copy of unique packages that are installed in other environments?

**Comment by kalefranz** [comment]
@ilanschnell did you ever push a PR for the offline repo/cache we discussed?

**Comment by hxu** [comment]
I'm trying to run the script, but it doesn't seem to write the index correctly (there's also a typo [here](https://github.com/ilanschnell/misc/blob/master/conda/c2r.py#L29) in pkgs_dirs), not sure why but trying to debug it now.

Just to make sure that I understand the intent of the script correctly, assuming a fresh install of Anaconda, are you supposed:
  - Install Anaconda
  - Use the root environment to run this script
  - This should create a channel folder with all of the packages that are installed by default with Anaconda
  - You can then create environments and install packages using `conda install -c <PATH_TO_FOLDER> --offline <PACKAGES>`

**Comment by hxu** [comment]
After reading the script in detail, it seems like there's a few things missing:

  - My `pkgs` directory in Anaconda 2.5.0 is unpacked, so I think you'd have to first zip up any packages that are found.  The test to determine if a folder is a package that I came up with is look to ensure that the `info/index.json` file exists.
  - I don't really understand why you're reading the `cache/*.json` files -- I presume these are stored from the last time that conda called out to a real channel
  - You could probably just use `conda index` from the conda-build package to do the indexing: https://github.com/conda/conda-build/blob/master/conda_build/index.py

**Comment by ijstokes** [comment]
I did a stack of experimentation with this earlier today and have now lost my terminal session with traces to share.  But they were motivated by #1597, and basically I came to the same conclusion: I couldn't get `--offline` to work for me.

**Comment by bgruening** [comment]
I end up in a `RuntimeError: maximum recursion depth exceeded` error with the `--offline` argument.

The problem exists since conda 4.0. I was able to get it running with "conda install conda=3.19.0"

```
 % conda --version
conda 4.0.6
```

Working command is:
```
conda --debug create -y --unknown --file packages.txt -c bioconda
```

Broken command is:
```
conda --debug create -y --unknown  --offline --file packages.txt -c bioconda
```

packages.txt:
```
# This file may be used to create an environment using:
# $ conda create --name <env> --file <this file>
# platform: linux-64
libgcc=5.2.0=0
libxml2=2.9.3=0
openbabel=2.3.90dev7d621d9=py27_0
openssl=1.0.2h=0
pip=8.1.1=py27_1
python=2.7.11=0
readline=6.2=2
setuptools=20.7.0=py27_0
sqlite=3.9.2=0
tk=8.5.18=0
wheel=0.29.0=py27_0
zlib=1.2.8=0
```

**Comment by bgruening** [comment]
Any news on this bug? Can I help to track it down?

**Comment by kalefranz** [comment]
Offline bugs are iteratively being quashed in 4.1.x.  Any changes to the package cache are now slated for 4.3 or later.

**Comment by bgruening** [comment]
@kalefranz it's getting worse :(
Because of this bug we are stucked to use conda `3.19.3` but it turns out with this older version I'm not able to install newer versions of some packages, only older ones.

xref: https://github.com/bgruening/galaxytools/pull/390

**Comment by bgruening** [comment]
@kalefranz is there anything we can do to get this fixed or should we remove the `--offline` option?

**Comment by kalefranz** [comment]
This involves package cache issues which will be handled in 4.4.

**Comment by bgruening** [comment]
@kalefranz any estimate for 4.4? This bug is really pressing for the Galaxy community. We are stuck with version 3 :(

**Comment by HassanAmr** [comment]
Is there any update on this issue? This is really important for us! Thanks!

**Comment by kalefranz** [comment]
The first canary release of `4.3.` will hopefully be published on or before Dec. 1.  Then we'll start working on `4.4` :)

**Comment by kalefranz** [comment]
Fixed in 4.3.
